marshal/consumer.go

```diff
@@ -375,6 +394,35 @@ func (c *Consumer) claimTopics() {
 		}
 	}
 
+	// let's compare the new topic claims vs old ones. Upon change, we should trigger an update
+	c.lock.RLock()
+	defer c.lock.RUnlock()
+	changed := false
+	// check for missing topic claims
+	for topic := range c.claimedTopics {
+		if !latestClaims[topic] {
+			changed = true
+			break
+		}
+	}
+	// check for new topic claims
+	if !changed && len(latestClaims) != len(c.claimedTopics) {
+		changed = true
+	}
+
+	if changed {
+		c.lock.RUnlock()
```

Comment:
This is already not serial isn't it? You can't atomically upgrade a read lock to write, so as soon as you release on line 426 you have no idea when this writer function is going to run and how much has happened between then and now. So I don't understand what you mean when you say it is easier when synchronous because it is already asynchronous as you wrote it.

Releasing a lock you didn't acquire is really ugly and is hard to reason about even if there is only one caller. Deferring to lock is bananas :)

Why not just have this function acquire the RLock it needs and then call a helper than acquires the Lock it needs?